File: templates/templates/php/Tests/primary/AuthAdditionalTest.php
declare(strict_types=1);

namespace OpenAPI\OpenAPI\Tests;

use OpenAPI\OpenAPI\Tests\CommonHelpers;
use PHPUnit\Framework\TestCase;

final class AuthAdditionalTest extends TestCase
{
    public function testGlobalSecurityFieldsOrdering(): void
    {
        CommonHelpers::recordTest('auth-global-security-fields-ordering');

        // When `maintainOpenApiOrder: false` is set, security fields are sorted in alphabetical order.
        // Since the first non-nil field is selected ApiKeyAuth takes precedence over BasicHttp,
        // which is not valid authentication for this endpoint.
        $security = new \OpenAPI\OpenAPI\Models\Shared\Security(
            apiKeyAuth: 'testApiKey',
            basicHttp: new \OpenAPI\OpenAPI\Models\Shared\SchemeBasicHTTP(
                username: 'testUser',
                password: 'testPass',
            ),
        );
        $sdk = \OpenAPI\OpenAPI\SDK::builder()->setSecurity($security)->build();

        try {
            $sdk->auth->globalSecurityBasicHttp();
            $this->fail('Expected APIException was not thrown');
        } catch (\OpenAPI\OpenAPI\Models\Errors\APIException $e) {
            $this->assertEquals(401, $e->getCode());
        }
    }

    public function testHoistedSecurityAccessTokenOnly(): void
    {
        CommonHelpers::recordTest('auth-hoisted-security-access-token-only');

        $security = new \OpenAPI\OpenAPI\Models\Shared\Security(
            apiKeyAuth: 'testApiKey',
            basicHttp: new \OpenAPI\OpenAPI\Models\Shared\SchemeBasicHTTP(
                username: 'testUser',
                password: 'testPass',
            ),
            accessToken: 'Bearer ghp_xxxx',
        );
        $sdk = \OpenAPI\OpenAPI\SDK::builder()->setSecurity($security)->build();

        $response = $sdk->auth->hoistedSecurityAccessTokenOnly();

        $this->assertNotNull($response);
        $this->assertEquals(200, $response->httpMeta->response->getStatusCode());
        $this->assertNotNull($response->token);
        $this->assertEquals('Bearer ghp_xxxx', $response->token->token);
    }

    public function testHoistedSecurityAccessTokenFirst(): void
    {
        CommonHelpers::recordTest('auth-hoisted-security-access-token-first');

        $security = new \OpenAPI\OpenAPI\Models\Shared\Security(
            apiKeyAuth: 'testApiKey',
            accessToken: 'Bearer ghp_xxxx',
        );
        $sdk = \OpenAPI\OpenAPI\SDK::builder()->setSecurity($security)->build();

        $response = $sdk->auth->hoistedSecurityAccessTokenFirst();

        $this->assertNotNull($response);
        $this->assertEquals(200, $response->httpMeta->response->getStatusCode());
        $this->assertNotNull($response->token);
        $this->assertEquals('Bearer ghp_xxxx', $response->token->token);
    }

    public function testHoistedSecurityApiKeyFirst(): void
    {
        CommonHelpers::recordTest('auth-hoisted-security-api-key-first');

        $security = new \OpenAPI\OpenAPI\Models\Shared\Security(
            apiKeyAuth: 'testApiKey',
            accessToken: 'Bearer ghp_xxxx',
        );
        $sdk = \OpenAPI\OpenAPI\SDK::builder()->setSecurity($security)->build();

        $response = $sdk->auth->hoistedSecurityApiKeyFirst();

        $this->assertNotNull($response);
        $this->assertEquals(200, $response->httpMeta->response->getStatusCode());
        $this->assertNotNull($response->token);
        $this->assertEquals('testApiKey', $response->token->token);
    }

    public function testHoistedSecurityBasicHttpOnly(): void
    {
        CommonHelpers::recordTest('auth-hoisted-security-basic-http-only');

        $security = new \OpenAPI\OpenAPI\Models\Shared\Security(
            basicHttp: new \OpenAPI\OpenAPI\Models\Shared\SchemeBasicHTTP(
                username: 'testUser',
                password: 'testPass',
            ),
            apiKeyAuth: 'testApiKey',
            accessToken: 'Bearer ghp_xxxx',
        );
        $sdk = \OpenAPI\OpenAPI\SDK::builder()->setSecurity($security)->build();

        $response = $sdk->auth->hoistedSecurityBasicHttpOnly();

        $this->assertNotNull($response);
        $this->assertEquals(200, $response->httpMeta->response->getStatusCode());
        $this->assertNotNull($response->basicAuth);
        $this->assertTrue($response->basicAuth->authenticated);
        $this->assertEquals('testUser', $response->basicAuth->user);
    }

    public function testHoistedSecurityInvalidField(): void
    {
        CommonHelpers::recordTest('auth-hoisted-security-invalid-field');

        // Provide only BasicHttp — not valid for accessTokenFirst which expects bearer/apiKey
        $security = new \OpenAPI\OpenAPI\Models\Shared\Security(
            basicHttp: new \OpenAPI\OpenAPI\Models\Shared\SchemeBasicHTTP(
                username: 'user',
                password: 'pass',
            ),
        );
        $sdk = \OpenAPI\OpenAPI\SDK::builder()->setSecurity($security)->build();

        try {
            $sdk->auth->hoistedSecurityAccessTokenFirst();
            $this->fail('Expected APIException was not thrown');
        } catch (\OpenAPI\OpenAPI\Models\Errors\APIException $e) {
            $this->assertEquals(401, $e->getCode());
        }
    }
}

File: templates/templates/ruby/tests/primary/auth_additional_test.rb
# frozen_string_literal: true
# typed: true

require_relative "../lib/openapi"
require_relative "common_helper_test"
require_relative "helper_test"

require "minitest/autorun"
require "minitest/focus"
require "rack"

module OpenApiSDK
  class TestGlobalSecurityFieldsOrdering < Minitest::Test
    def test_global_security_fields_ordering
      record_test("auth-global-security-fields-ordering")

      # When maintainOpenApiOrder: false is set, the first non-nil field (in alphabetical order) is selected.
      # In this case api_key_auth takes precedence over basic_http, which is invalid for this endpoint.
      sdk = OpenApiSDK::SDK.new(
        server_url: CommonHelpers.httpbin_url,
        security: Models::Shared::Security.new(
          api_key_auth: "testApiKey",
          basic_http: Models::Shared::SchemeBasicHTTP.new(
            username: "testUser",
            password: "testPass"
          )
        )
      )

      err = assert_raises(Models::Errors::APIError) do
        sdk.auth.global_security_basic_http
      end
      assert_equal(401, err.status_code)
    end
  end

  class TestHoistedSecurityAccessTokenOnly < Minitest::Test
    def test_hoisted_security_access_token_only
      record_test("auth-hoisted-security-access-token-only")

      sdk = OpenApiSDK::SDK.new(
        server_url: CommonHelpers.httpbin_url,
        security: Models::Shared::Security.new(
          api_key_auth: "testApiKey",
          basic_http: Models::Shared::SchemeBasicHTTP.new(
            username: "testUser",
            password: "testPass"
          ),
          access_token: "Bearer ghp_xxxx"
        )
      )

      res = sdk.auth.hoisted_security_access_token_only

      refute_nil(res)
      assert_equal(Rack::Utils.status_code(:ok), res.http_meta.response.status)
      refute_nil(res.token)
      assert_equal("Bearer ghp_xxxx", res.token.token)
    end
  end

  class TestHoistedSecurityAccessTokenFirst < Minitest::Test
    def test_hoisted_security_access_token_first
      record_test("auth-hoisted-security-access-token-first")

      sdk = OpenApiSDK::SDK.new(
        server_url: CommonHelpers.httpbin_url,
        security: Models::Shared::Security.new(
          api_key_auth: "testApiKey",
          access_token: "Bearer ghp_xxxx"
        )
      )

      res = sdk.auth.hoisted_security_access_token_first

      refute_nil(res)
      assert_equal(Rack::Utils.status_code(:ok), res.http_meta.response.status)
      refute_nil(res.token)
      assert_equal("Bearer ghp_xxxx", res.token.token)
    end
  end

  class TestHoistedSecurityApiKeyFirst < Minitest::Test
    def test_hoisted_security_api_key_first
      record_test("auth-hoisted-security-api-key-first")

      sdk = OpenApiSDK::SDK.new(
        server_url: CommonHelpers.httpbin_url,
        security: Models::Shared::Security.new(
          api_key_auth: "testApiKey",
          access_token: "Bearer ghp_xxxx"
        )
      )

      res = sdk.auth.hoisted_security_api_key_first

      refute_nil(res)
      assert_equal(Rack::Utils.status_code(:ok), res.http_meta.response.status)
      refute_nil(res.token)
      assert_equal("testApiKey", res.token.token)
    end
  end

  class TestHoistedSecurityBasicHttpOnly < Minitest::Test
    def test_hoisted_security_basic_http_only
      record_test("auth-hoisted-security-basic-http-only")

      sdk = OpenApiSDK::SDK.new(
        server_url: CommonHelpers.httpbin_url,
        security: Models::Shared::Security.new(
          api_key_auth: "testApiKey",
          basic_http: Models::Shared::SchemeBasicHTTP.new(
            username: "testUser",
            password: "testPass"
          ),
          access_token: "Bearer ghp_xxxx"
        )
      )

      res = sdk.auth.hoisted_security_basic_http_only

      refute_nil(res)
      assert_equal(Rack::Utils.status_code(:ok), res.http_meta.response.status)
      refute_nil(res.basic_auth)
      assert(res.basic_auth.authenticated)
      assert_equal("testUser", res.basic_auth.user)
    end
  end

  class TestHoistedSecurityInvalidField < Minitest::Test
    def test_hoisted_security_invalid_field
      record_test("auth-hoisted-security-invalid-field")

      # Provide only basic_http — not valid for access_token_first which expects bearer/apiKey
      sdk = OpenApiSDK::SDK.new(
        server_url: CommonHelpers.httpbin_url,
        security: Models::Shared::Security.new(
          basic_http: Models::Shared::SchemeBasicHTTP.new(
            username: "user",
            password: "pass"
          )
        )
      )

      err = assert_raises(Models::Errors::APIError) do
        sdk.auth.hoisted_security_access_token_first
      end
      assert_equal(401, err.status_code)
    end
  end
end
